Mathlib/Data/MLList/BestFirst.lean

```diff
@@ -29,37 +31,198 @@ Options:
+except in cases where the node arrives at the front of the queue.
+-/
+section
+
+/-- A node in a `BestFirstQueue. -/
+structure BestFirstNode (prio : α → Thunk ω) (ε : α → Type) where
+  /-- The data to store at a node, from which we can calculate a priority using `prio`. -/
+  key : α
+  /-- An estimator for the priority of the key.
+  (We will assume we have `[∀ a : α, Estimator (prio a) (ε a)]`.) -/
+  estimator : ε key
+
+variable {α : Type} {prio : α → Thunk ω} {ε : α → Type} [LinearOrder ω]
+  [∀ a, Estimator (prio a) (ε a)]
+  [∀ a : α, WellFoundedGT (range (bound (prio a) : ε a → ω))]
+  {m : Type → Type} [Monad m] {β : Type}
+
+/-- Calculate the current best lower bound for the the priority of a node. -/
+def BestFirstNode.estimate (n : BestFirstNode prio ε) : ω := bound (prio n.key) n.estimator
+
+instance [Ord ω] [Ord α] : Ord (BestFirstNode prio ε) where
+  compare :=
+    compareLex
+      (compareOn BestFirstNode.estimate)
+      (compareOn BestFirstNode.key)
+
+set_option linter.unusedVariables false in
+variable (prio ε m β) [Ord ω] [Ord α] in
+/-- A queue of `MLList m β`s, lazily prioritized by lower bounds. -/
+@[nolint unusedArguments]
+def BestFirstQueue (maxSize : Option Nat) := RBMap (BestFirstNode prio ε) (MLList m β) compare
+
+variable [Ord ω] [Ord α] {maxSize : Option Nat}
+
+namespace BestFirstQueue
+
+/--
+Add a new `MLList m β` to the `BestFirstQueue`, and if this takes the size above `maxSize`,
+eject a `MLList` from the tail of the queue.
+-/
+-- Note this ejects the element with the greatest estimated priority,
+-- not necessarily the greatest priority!
+def insertAndEject
+    (q : BestFirstQueue prio ε m β maxSize) (n : BestFirstNode prio ε) (l : MLList m β) :
+    BestFirstQueue prio ε m β maxSize :=
+  match maxSize with
+  | none => q.insert n l
+  | some max =>
+    if q.size < max then
+      q.insert n l
+    else
+      match q.max with
+      | none => RBMap.empty
+      | some m => q.insert n l |>.erase m.1
+
+/--
+Pop a `β` off the `MLList m β` with lowest priority,
+also returning the index in `α` and the current best lower bound for its priority.
+This may require improving estimates of priorities and shuffling the queue.
+-/
+partial def popWithBound (q : BestFirstQueue prio ε m β maxSize) :
+    m (Option (((a : α) × (ε a) × β) × BestFirstQueue prio ε m β maxSize)) := do
+  let s := @toStream (RBMap _ _ _) _ _ q
+  match s.next? with
+  | none => pure none
+  | some ((n, l), s') => match s'.next? with
+    | none => do match ← l.uncons with
+      | none => pure none
+      | some (b, l') => pure <| some (⟨n.key, n.estimator, b⟩, RBMap.single n l')
+    | some ((m, _), _) =>
+      match improveUntil (prio n.key) (m.estimate < ·) n.estimator with
+      | .error e => do match ← l.uncons with
+        | none => popWithBound (q.erase n)
+        | some (b, l') =>
+          match e with
+          | none => pure <| some (⟨n.key, n.estimator, b⟩, q.modify n fun _ => l')
```

Comment:
One thing I'm noticing is that the control flow is quite deeply nested. If we weren't already in a monadic context, many `match` statements on `Option`s could be replaced with binds:
```
  match s.next? with
  | none => pure none
  | some ((n, l), s') => ...
```
could morally become:
```
  let ((n, l), s') ← s.next?
```
Maybe there is still a way to do this with Lean's fancier `do`?

This also reappears a few times below.

Reply:
Regardless of monadology, how about moving the logic that improves the estimate for the queue's head to a separate definition? I imagine it could also be useful in other applications (for example, peek).

Reply:
I didn't have any luck with the monadology, but the refactoring suggestion was a great improvement. I've also added lots of comments explaining the flow control.